src/main/java/io/split/android/client/impressions/ImpressionsStorageManager.java

```diff
@@ -175,13 +260,95 @@ private void loadImpressionsFromDisk(){
+                String fileName = String.format(IMPRESSIONS_FILE_NAME, i);
+                mFileStorageManager.write(fileName, json);
+                i++;
+            } catch (IOException e) {
+                Logger.e(e, "Could not save impressions");
+            } catch (JsonSyntaxException syntaxException) {
+                Logger.e(syntaxException, "Unable to parse impressions to save");
+            }
+        }
+    }
+
+    private List<ChunkHeader> getChunkHeaders(Map<String, StoredImpressions> storedImpressions) {
+        List<ChunkHeader> chunkHeaders = new ArrayList<>();
+        for (StoredImpressions storedImpression : storedImpressions.values()) {
+            ChunkHeader header = new ChunkHeader(storedImpression.id(), storedImpression.getAttempts(), storedImpression.getTimestamp());
+            chunkHeaders.add(header);
+        }
+        return chunkHeaders;
+    }
+
+    private List<Map<String, List<KeyImpression>>> splitChunks(List<StoredImpressions> storedImpressions) {
+
+        List<Map<String, List<KeyImpression>>> splitImpressions = new ArrayList<>();
+        long bytesCount = 0;
+        List<KeyImpression> currentImpressions = new ArrayList<>();
+        Map<String, List<KeyImpression>> currentChunk = new HashMap<>();
+        for (StoredImpressions storedImpression : storedImpressions) {
+            List<TestImpressions> testImpressions = storedImpression.impressions();
+            for (TestImpressions testImpression : testImpressions) {
+                List<KeyImpression> keyImpressions = testImpression.keyImpressions;
+                for (KeyImpression keyImpression : keyImpressions) {
+                    if (bytesCount + ESTIMATED_IMPRESSION_SIZE > MAX_BYTES_PER_CHUNK) {
+                        currentChunk.put(buildImpressionRecordKey(storedImpression.id(), testImpression.testName), currentImpressions);
+                        splitImpressions.add(currentChunk);
+                        currentChunk = new HashMap<>();
+                        currentImpressions = new ArrayList<>();
+                        bytesCount = 0;
+                    }
+                    currentImpressions.add(keyImpression);
+                    bytesCount += ESTIMATED_IMPRESSION_SIZE;
+                }
+                if (currentImpressions.size() > 0) {
+                    currentChunk.put(buildImpressionRecordKey(storedImpression.id(), testImpression.testName), currentImpressions);
+                    currentImpressions = new ArrayList<>();
+                }
+            }
+        }
+        splitImpressions.add(currentChunk);
+        return splitImpressions;
+    }
+
+    private String buildImpressionRecordKey(String storedImpressionId, String testName) {
+        return String.format("%s_%s", storedImpressionId, testName);
+    }
+
+    private String getStoredImpressionsIdFromRecordKey(String recordKey) {
+        return recordKey.substring(0, recordKey.indexOf("_"));
```

Comment:
can this throw an out of index exception in any scenario?

Reply:
Yup, right. Added exception handling.